beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java

```diff
@@ -0,0 +1,194 @@
+
+    private final String name;
+    private final long limit;
+    private final RepositoryFactory repositoryFactory;
+    private final CommitLogReader.Factory<?> readerFactory;
+
+    private transient boolean externalizableOffsets = false;
+    private transient BlockingQueueLogObserver observer;
+
+    private CommitLogReadFn(
+        String name,
+        long limit,
+        RepositoryFactory repositoryFactory,
+        CommitLogReader.Factory<?> readerFactory) {
+
+      this.name = name;
+      this.limit = limit;
+      this.repositoryFactory = repositoryFactory;
+      this.readerFactory = readerFactory;
+    }
+
+    @ProcessElement
+    public ProcessContinuation process(
+        RestrictionTracker<OffsetRestrictionTracker.StartEndOffset, Offset> tracker,
+        OutputReceiver<StreamElement> output) {
+
+      try (ObserveHandle handle = startObserve(tracker.currentRestriction())) {
+        int i = 0;
+        while (!Thread.currentThread().isInterrupted()
+            && observer.getWatermark() < Watermarks.MAX_WATERMARK) {
+
+          StreamElement element = observer.take();
+          if (element == null) {
+            ExceptionUtils.ignoringInterrupted(() -> TimeUnit.MILLISECONDS.sleep(100));
+            continue;
+          }
+          boolean isFirst = i++ == 0;
+          Offset offset = Objects.requireNonNull(observer.getLastReadContext().getOffset());
+          if (externalizableOffsets
+              && !tracker.currentRestriction().isStartInclusive()
+              && isFirst) {
+            // skip the exclusive first offset
+            continue;
+          }
+          if (!tracker.tryClaim(offset)) {
+            return ProcessContinuation.stop();
+          }
+          output.outputWithTimestamp(element, Instant.ofEpochMilli(element.getStamp()));
+        }
+      }
+      boolean terminated = observer.getWatermark() >= Watermarks.MAX_WATERMARK;
+      return terminated
+          ? ProcessContinuation.stop()
+          : ProcessContinuation.resume().withResumeDelay(Duration.millis(100));
+    }
+
+    private ObserveHandle startObserve(OffsetRestrictionTracker.StartEndOffset restriction) {
+      CommitLogReader reader = readerFactory.apply(repositoryFactory.apply());
+      this.externalizableOffsets = reader.hasExternalizableOffsets();
+      observer = BlockingQueueLogObserver.create(name, limit, Watermarks.MIN_WATERMARK);
+      return reader.observeBulkOffsets(
+          Collections.singletonList(restriction.getStartOffset()), observer);
+    }
+
+    @GetInitialRestriction
+    public OffsetRestrictionTracker.StartEndOffset initialRestriction() {
+      return OffsetRestrictionTracker.StartEndOffset.initialRestriction();
+    }
+
+    @SplitRestriction
+    public void splitRestriction(
+        @Restriction OffsetRestrictionTracker.StartEndOffset restriction,
+        OutputReceiver<OffsetRestrictionTracker.StartEndOffset> splits) {
```

Comment:
what happens if we produce more splits than configured parallelism? can runner handle it the same way as in UnboindedSourceWrapper?

Reply:
Definitely has to. There is no parameter telling you what the parallelism is.